# Ebay used cars dataset

For this exercise, I am going to work on cleaning the data so that it can be easier to use. 
Data was derived from Kaggle: https://www.kaggle.com/orgesleka/used-cars-database/data

### Data dictionary

|Data type |Meaning|
|----------|-------|
|dateCrawled | When this ad was first crawled. All field-values are taken from this date. |
|name | Name of the car.|
|seller | Whether the seller is private or a dealer.|
|offerType | The type of listing. |
|price | The price on the ad to sell the car.|
|abtest | Whether the listing is included in an A/B test.|
|vehicleType | The vehicle Type. |
|yearOfRegistration | The year in which the car was first registered.|
|gearbox | The transmission type. |
|powerPS | The power of the car in PS. |
|model | The car model name. |
|kilometer | How many kilometers the car has driven. |
|monthOfRegistration | The month in which the car was first registered.|
|fuelType | What type of fuel the car uses.|
|brand | The brand of the car.|
|notRepairedDamage | If the car has a damage which is not yet repaired. |
|dateCreated | The date on which the eBay listing was created.|
|nrOfPictures | The number of pictures in the ad.|
|postalCode | The postal code for the location of the vehicle.|
|lastSeenOnline | When the crawler saw this ad last online. |



In [2]:
import pandas as pd

In [3]:
#When you use UTF-8 encoding, I get an error. Decided to use Latin-1 and this brought no error

autos= pd.read_csv("autos.csv", encoding="Latin-1")

In [4]:
# get a snippet of what is in the actual data

autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

In [5]:
#Learning more about the dataframe. 
print(autos.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

In [6]:
print(autos.head())

           dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere 

## Things to note: 

1. Date created and date crawled have both date and time together and are not in date format. 
2. Year of registration is <int> instead of <date>
3. Price and odometer should be changes to integers and not string and the column name updated. 
4. Some columns are missing values: Vehicle type. Gear box, Model, Fuel type, Not repaired damage
5. The name column can be cleaner

### Step 1: Rename the columns

In [7]:
#print an array of existing column names

autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [8]:
#Make a copy of the column array

copy2=autos.columns.copy()

In [9]:
# Confirm everything is okay

copy2

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [10]:
# making edits to the columns through creating a data dictionary
# Change the rest of the columnn names from camelcase to snakecase

new_columns= {
   'dateCrawled': 'date_crawled',
    'name': 'name', 
    'seller': 'seller',
    'offerType': 'offer_type', 
    'price': 'price', 
    'abtest': 'abtest',
    'vehicleType': 'vehicle_type',
    'yearOfRegistration': 'registration_year', 
    'gearbox': 'gearbox', 
    'powerPS': 'power_ps',
    'model': 'model',
    'odometer': 'odometer',
    'monthOfRegistration': 'registration_month',
    'fuelType': 'fuel_type',
    'brand': 'brand',
    'notRepairedDamage': 'unrepaired_damage', 
    'dateCreated': 'ad_created', 
    'nrOfPictures':'nr_of_pictures',
    'postalCode':'postal_code',
    'lastSeen':'last_seen'   
}

In [11]:
#Assign the modified column names (copy2) back to the autos.columns attribute
columns = autos.columns

#change this into a series because map cannot be used on dict. 

autos.columns = pd.Series(columns).map(new_columns)

In [12]:
#let's print and see: 

autos.head(3)

date_crawled                                        name  seller  \
0  2016-03-26 17:47:46            Peugeot_807_160_NAVTECH_ON_BOARD  privat   
1  2016-04-04 13:38:56  BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  privat   
2  2016-03-26 18:57:24                  Volkswagen_Golf_1.6_United  privat   

  offer_type   price   abtest vehicle_type  registration_year    gearbox  \
0    Angebot  $5,000  control          bus               2004    manuell   
1    Angebot  $8,500  control    limousine               1997  automatik   
2    Angebot  $8,990     test    limousine               2009    manuell   

   power_ps   model   odometer  registration_month fuel_type       brand  \
0       158  andere  150,000km                   3       lpg     peugeot   
1       286     7er  150,000km                   6    benzin         bmw   
2       102    golf   70,000km                   7    benzin  volkswagen   

  unrepaired_damage           ad_created  nr_of_pictures  postal_code  \
0              nein  2016-03-26 00:00:00               0        79588   
1              nein  2016-04-04 00:00:00               0        71034   
2              nein  2016-03-26 00:00:00               0        35394   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37

### Step 2: More data exploration to see what else needs to be cleaned up


In [13]:
# For descriptive statistics add include all to get both categorical and numerical data

autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-30 17:37:35  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

#### Things to note and possibly change: 

1. Columns with missing values: vehicle_type, gearbox, model, fuel_type, unrepaired_damage
2. Columns with mostly one value which can be dropped: Seller, offer-type
3. Columns with numerical data stored as string: price, odometer,nr_of_pictures

###  Step 3: Change price and odometer values to numeric

In [14]:
#Because these are numerical values and for wasy readability, I will specifically say that I want the numbers to be displayed in this format

pd.set_option('float_format', '{:f}'.format)

In [15]:
# You can use multiple lines

autos["price"] = autos["price"].str.replace("$", "")
autos["price"] = autos["price"].str.replace(",", "")
autos["price"] = autos["price"].astype(int)

In [16]:
autos.price

0         5000
1         8500
2         8990
3         4350
4         1350
5         7900
6          300
7         1990
8          250
9          590
10         999
11         350
12        5299
13        1350
14        3999
15       18900
16         350
17        5500
18         300
19        4150
20        3500
21       41500
22       25450
23        7999
24       48500
25          90
26         777
27           0
28        5250
29        4999
         ...  
49970    15800
49971      950
49972     3300
49973     6000
49974        0
49975     9700
49976     5900
49977     5500
49978      900
49979    11000
49980      400
49981     2000
49982     1950
49983      600
49984        0
49985     1000
49986    15900
49987    21990
49988     9550
49989      150
49990    17500
49991      500
49992     4800
49993     1650
49994     5000
49995    24900
49996     1980
49997    13200
49998    22900
49999     1250
Name: price, Length: 50000, dtype: int64

In [17]:
# Or condense this into one readable code. 

autos["odometer"]= autos["odometer"].str.replace(",","").str.replace("km","").astype(int)

In [18]:
autos.odometer

0        150000
1        150000
2         70000
3         70000
4        150000
5        150000
6        150000
7        150000
8        150000
9        150000
10       150000
11       150000
12        50000
13       150000
14       150000
15        80000
16       150000
17       150000
18       150000
19       150000
20       150000
21       150000
22        10000
23       150000
24        30000
25       150000
26       125000
27       150000
28       150000
29       150000
          ...  
49970     60000
49971    150000
49972    150000
49973    150000
49974    150000
49975    100000
49976    150000
49977    150000
49978    150000
49979     70000
49980    125000
49981    150000
49982     90000
49983    150000
49984    150000
49985    150000
49986    125000
49987     50000
49988    150000
49989    150000
49990     30000
49991    150000
49992    125000
49993    150000
49994    150000
49995    100000
49996    150000
49997      5000
49998     40000
49999    150000
Name: odometer, Length: 

In [19]:
#An alternative way would be to create a function to do this. 

def cleaner(entry): 
    replacables = ["$", ",", "km", "KM", "Km"]
    for r in replacables: 
        entry = entry.replace(r, "")
        
    entry=entry.pd.to_numeric(entry)
    return entry

### Step 4: Exploration in search of outliers

In [20]:
#to view min/max/median/mean etc

autos.odometer.describe()

count    50000.000000
mean    125732.700000
std      40042.211706
min       5000.000000
25%     125000.000000
50%     150000.000000
75%     150000.000000
max     150000.000000
Name: odometer, dtype: float64

In [21]:
# to see how many unique values

autos.odometer.unique().shape

(13,)

In [24]:
#the maximum price looks a bit too much. Let's use values_count to see why so. 

autos.odometer.value_counts()

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer, dtype: int64

In [22]:
#to view min/max/median/mean etc

autos.price.describe()

count      50000.000000
mean        9840.043760
std       481104.380500
min            0.000000
25%         1100.000000
50%         2950.000000
75%         7200.000000
max     99999999.000000
Name: price, dtype: float64

In [23]:
#to see how many unique values

autos.price.unique().shape

(2357,)

In [25]:
autos.price.value_counts()

0         1421
500        781
1500       734
2500       643
1000       639
1200       639
600        531
800        498
3500       498
2000       460
999        434
750        433
900        420
650        419
850        410
700        395
4500       394
300        384
2200       382
950        379
1100       376
1300       371
3000       365
550        356
1800       355
5500       340
1250       335
350        335
1600       327
1999       322
          ... 
46200        1
29600        1
13480        1
21700        1
7373         1
3279         1
4286         1
188          1
17830        1
9130         1
910          1
238          1
2671         1
69900        1
151990       1
2479         1
4510         1
86500        1
47499        1
16998        1
27299        1
41850        1
4780         1
686          1
6495         1
20790        1
8970         1
846          1
2895         1
33980        1
Name: price, Length: 2357, dtype: int64

### This shows that majority of the cars cost between 500 - 5,500 USD. Implying that the car costing 99,999,999 USD is a possible outlier. 
*Also This is also ebay and these are 'used' cars so logically, this makes no sense* 

*According to https://interestingengineering.com/top-10-most-expensive-car-auctions the most expensive Second hand car to be sold was a Ferrari that cost 48 Million dollars). Even though the chances of selling a used Ferrari on Ebay are slim, we will use this as a benchmark to weed off outliers*

In [26]:
#Removing the outliers from the data and re-assigning this to the dataframe

autos = autos[autos["price"].between(1,48000000)]

In [27]:
#Checking to see how different the descriptive statistics are: 

autos.describe()

price  registration_year     power_ps      odometer  \
count    48578.000000       48578.000000 48578.000000  48578.000000   
mean      8069.541541        2004.753119   117.209869 125765.675820   
std     180035.634653          88.632571   200.652070  39793.193823   
min          1.000000        1000.000000     0.000000   5000.000000   
25%       1200.000000        1999.000000    71.000000 125000.000000   
50%       3000.000000        2004.000000   107.000000 150000.000000   
75%       7490.000000        2008.000000   150.000000 150000.000000   
max   27322222.000000        9999.000000 17700.000000 150000.000000   

       registration_month  nr_of_pictures  postal_code  
count        48578.000000    48578.000000 48578.000000  
mean             5.781794        0.000000 50976.081230  
std              3.685768        0.000000 25749.448131  
min              0.000000        0.000000  1067.000000  
25%              3.000000        0.000000 30657.000000  
50%              6.000000        0.000000 49716.000000  
75%              9.000000        0.000000 71665.000000  
max             12.000000        0.000000 99998.000000

####  ✅ Much improvement from initial descriptive stats in [output 22]

### Step 5: Dealing with dates

In [28]:
#The columns we will be working with

autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [31]:
# the first 10 characters represent the day (e.g. 2016-03-12). To understand the date range, we can extract just the date values, use Series.value_counts() to generate a distribution, and then sort by the index.

print(autos['date_crawled'].str[:10].head())

0    2016-03-26
1    2016-04-04
2    2016-03-26
3    2016-03-12
4    2016-04-01
Name: date_crawled, dtype: object


In [32]:
#Shows the distribution across the dates. 

autos['date_crawled'].str[:10].value_counts()

2016-04-03    1875
2016-03-20    1840
2016-03-21    1817
2016-03-12    1794
2016-03-14    1775
2016-04-04    1774
2016-03-07    1749
2016-04-02    1723
2016-03-28    1693
2016-03-19    1689
2016-03-15    1665
2016-03-29    1658
2016-03-30    1636
2016-04-01    1636
2016-03-08    1619
2016-03-09    1608
2016-03-22    1603
2016-03-11    1582
2016-03-23    1565
2016-03-26    1564
2016-03-10    1563
2016-03-31    1547
2016-03-17    1537
2016-03-25    1535
2016-03-27    1510
2016-03-16    1438
2016-03-24    1425
2016-03-05    1230
2016-03-13     761
2016-03-06     682
2016-04-05     636
2016-03-18     627
2016-04-06     154
2016-04-07      68
Name: date_crawled, dtype: int64

In [39]:
#To include missing values in the distribution and to use percentages instead of counts, chain the Series.value_counts(normalize=True, dropna=False) method.

#1. For date_crawled
autos['date_crawled'].str[:7].value_counts(normalize=True, dropna=False).sort_index()

2016-03   0.838075
2016-04   0.161925
Name: date_crawled, dtype: float64

In [35]:
#2. For ad_created
autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2015-06-11   0.000021
2015-08-10   0.000021
2015-09-09   0.000021
2015-11-10   0.000021
2015-12-05   0.000021
2015-12-30   0.000021
2016-01-03   0.000021
2016-01-07   0.000021
2016-01-10   0.000041
2016-01-13   0.000021
2016-01-14   0.000021
2016-01-16   0.000021
2016-01-22   0.000021
2016-01-27   0.000062
2016-01-29   0.000021
2016-02-01   0.000021
2016-02-02   0.000041
2016-02-05   0.000041
2016-02-07   0.000021
2016-02-08   0.000021
2016-02-09   0.000021
2016-02-11   0.000021
2016-02-12   0.000041
2016-02-14   0.000041
2016-02-16   0.000021
2016-02-17   0.000021
2016-02-18   0.000041
2016-02-19   0.000062
2016-02-20   0.000041
2016-02-21   0.000062
               ...   
2016-03-09   0.033163
2016-03-10   0.031887
2016-03-11   0.032896
2016-03-12   0.036766
2016-03-13   0.017004
2016-03-14   0.035181
2016-03-15   0.034007
2016-03-16   0.030117
2016-03-17   0.031290
2016-03-18   0.013586
2016-03-19   0.033678
2016-03-20   0.037939
2016-03-21   0.037610
2016-03-22   0.032813
2016-03-23

In [37]:
#2.A What year were most ads created
autos['ad_created'].str[:4].value_counts(normalize=True, dropna=False).sort_index()

2015   0.000124
2016   0.999876
Name: ad_created, dtype: float64

In [38]:
#2.B What month in 2016 were most ads created
autos['ad_created'].str[:7].value_counts(normalize=True, dropna=False).sort_index()

2015-06   0.000021
2015-08   0.000021
2015-09   0.000021
2015-11   0.000021
2015-12   0.000041
2016-01   0.000247
2016-02   0.001256
2016-03   0.837498
2016-04   0.160875
Name: ad_created, dtype: float64

In [40]:
#3. For last_seen

autos['last_seen'].str[:7].value_counts(normalize=True, dropna=False).sort_index()

2016-03   0.424122
2016-04   0.575878
Name: last_seen, dtype: float64

### From this data, some apparent findings include:
- We had most number of ads created in March 2016! 
- March 2016 is also when most ads were crawled

In [41]:
autos['registration_year'].describe()

count   48578.000000
mean     2004.753119
std        88.632571
min      1000.000000
25%      1999.000000
50%      2004.000000
75%      2008.000000
max      9999.000000
Name: registration_year, dtype: float64

*FYI It is impossible to have to have registration year as 9999 or 1000 as we are in 2020 so we have to fix this*

In [42]:
autos = autos[autos["registration_year"].between(1900,2020)]

In [43]:
#Let's re-run and see what we have now 

autos['registration_year'].describe()

count   48558.000000
mean     2003.459554
std         7.571216
min      1910.000000
25%      1999.000000
50%      2004.000000
75%      2008.000000
max      2019.000000
Name: registration_year, dtype: float64

#### ✅ Makes much more sense though am not sure about the car registered in 1910. 

In [44]:
autos['registration_year'].value_counts(normalize='True',dropna='False').sort_index()

1910   0.000103
1927   0.000021
1929   0.000021
1931   0.000021
1934   0.000041
1937   0.000082
1938   0.000021
1939   0.000021
1941   0.000041
1943   0.000021
1948   0.000021
1950   0.000062
1951   0.000041
1952   0.000021
1953   0.000021
1954   0.000041
1955   0.000041
1956   0.000082
1957   0.000041
1958   0.000082
1959   0.000124
1960   0.000494
1961   0.000124
1962   0.000082
1963   0.000165
1964   0.000247
1965   0.000350
1966   0.000453
1967   0.000535
1968   0.000535
         ...   
1990   0.007146
1991   0.006981
1992   0.007640
1993   0.008752
1994   0.012954
1995   0.025289
1996   0.028275
1997   0.040179
1998   0.048663
1999   0.059681
2000   0.064994
2001   0.054306
2002   0.051197
2003   0.055583
2004   0.055665
2005   0.060464
2006   0.055006
2007   0.046892
2008   0.045616
2009   0.042959
2010   0.032724
2011   0.033424
2012   0.026978
2013   0.016537
2014   0.013674
2015   0.008073
2016   0.025125
2017   0.028687
2018   0.009700
2019   0.000041
Name: registration_year,

#### ✅ Majority of the used cars on Ebay were registered in 2005 and 2000 with the least surprisingly not being 1910! 

### Step 6: Aggregating brand data

The aim is to: 
- Identify the unique values we want to aggregate by
- Create an empty dictionary to store our aggregate data
- Loop over the unique values, and for each:
    - Subset the dataframe by the unique values
    - Calculate the mean of whichever column we're interested in
    - Assign the val/mean to the dict as k/v.

In [45]:
#Let's explore different variations across different car brands

autos['brand'].unique()

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'seat', 'renault', 'mercedes_benz', 'audi', 'sonstige_autos',
       'opel', 'mazda', 'porsche', 'mini', 'toyota', 'dacia', 'nissan',
       'jeep', 'saab', 'volvo', 'mitsubishi', 'jaguar', 'fiat', 'skoda',
       'subaru', 'kia', 'citroen', 'chevrolet', 'hyundai', 'honda',
       'daewoo', 'suzuki', 'trabant', 'land_rover', 'alfa_romeo', 'lada',
       'rover', 'daihatsu', 'lancia'], dtype=object)

In [57]:
the_brands=autos['brand'].value_counts().sort_values().index

In [60]:
#creating an empty disctionary to hold the aggregate data about mean_price
mean_price = {}

for item in the_brands:
    #Loop over your selected brands, and assign the mean price to the dictionary, with the brand name as the key.
    selected_rows = autos[autos["brand"]==item]
    item_mean= selected_rows['price'].mean()
    mean_price[item]=item_mean


In [61]:
mean_price

{'alfa_romeo': 4041.8442367601247,
 'audi': 9212.9306621881,
 'bmw': 8493.857251184834,
 'chevrolet': 6620.069090909091,
 'chrysler': 3519.3846153846152,
 'citroen': 43578.90962099125,
 'dacia': 5897.736434108527,
 'daewoo': 1079.342105263158,
 'daihatsu': 1628.4262295081967,
 'fiat': 12566.541567695962,
 'ford': 7305.141843971631,
 'honda': 4000.139175257732,
 'hyundai': 5371.792960662526,
 'jaguar': 11525.554054054053,
 'jeep': 11590.214953271028,
 'kia': 5889.298550724638,
 'lada': 2647.7241379310344,
 'lancia': 3182.0,
 'land_rover': 18934.272727272728,
 'mazda': 4059.059539918809,
 'mercedes_benz': 8526.623225806452,
 'mini': 10541.566985645934,
 'mitsubishi': 3386.911838790932,
 'nissan': 4669.3859649122805,
 'opel': 5281.321706161138,
 'peugeot': 3065.611888111888,
 'porsche': 45624.27177700349,
 'renault': 2431.195698924731,
 'rover': 1586.4923076923078,
 'saab': 3183.493670886076,
 'seat': 4320.168661588684,
 'skoda': 6353.544871794872,
 'smart': 3518.102305475504,
 'sonstige_

#### ✅ Most expensive brands: Audi, BMW, Fiat (?), Jaguar, Jeep, Landrover, Porsche

#### 😉 Least expensive brands:  Daewoo, Trabant

In [63]:
#creating an empty dictionary to hold the aggregate data about mean_mileage
mean_mileage = {}

for item in the_brands:
    #Loop over your selected brands, and assign the mean price to the dictionary, with the brand name as the key.
    selected_rows = autos[autos["brand"]==item]
    item_mean= selected_rows['odometer'].mean()
    mean_mileage[item]=item_mean

In [64]:
mean_mileage

{'alfa_romeo': 131043.6137071651,
 'audi': 129492.56238003839,
 'bmw': 132686.2559241706,
 'chevrolet': 99472.72727272728,
 'chrysler': 132366.86390532544,
 'citroen': 120043.73177842566,
 'dacia': 84728.68217054264,
 'daewoo': 121644.73684210527,
 'daihatsu': 115901.6393442623,
 'fiat': 117593.03246239113,
 'ford': 124327.71867612294,
 'honda': 123092.78350515464,
 'hyundai': 106718.4265010352,
 'jaguar': 124527.02702702703,
 'jeep': 127102.80373831776,
 'kia': 112739.13043478261,
 'lada': 85000.0,
 'lancia': 123090.90909090909,
 'land_rover': 118333.33333333333,
 'mazda': 124871.44790257105,
 'mercedes_benz': 130848.3870967742,
 'mini': 88899.52153110047,
 'mitsubishi': 126549.11838790932,
 'nissan': 118711.20107962213,
 'opel': 129455.92417061611,
 'peugeot': 127356.64335664336,
 'porsche': 96759.5818815331,
 'renault': 128062.36559139784,
 'rover': 138230.76923076922,
 'saab': 143670.88607594935,
 'seat': 121768.22633297062,
 'skoda': 110993.58974358975,
 'smart': 100511.5273775216

In [67]:
#We can combine the data from both series objects into a single dataframe (with a shared index) and display the dataframe directly.
#Step1: Price 

#With this method, the keys in the dictionary became the index in the series object

P = pd.Series(mean_price)

print(P)


alfa_romeo        4041.844237
audi              9212.930662
bmw               8493.857251
chevrolet         6620.069091
chrysler          3519.384615
citroen          43578.909621
dacia             5897.736434
daewoo            1079.342105
daihatsu          1628.426230
fiat             12566.541568
ford              7305.141844
honda             4000.139175
hyundai           5371.792961
jaguar           11525.554054
jeep             11590.214953
kia               5889.298551
lada              2647.724138
lancia            3182.000000
land_rover       18934.272727
mazda             4059.059540
mercedes_benz     8526.623226
mini             10541.566986
mitsubishi        3386.911839
nissan            4669.385965
opel              5281.321706
peugeot           3065.611888
porsche          45624.271777
renault           2431.195699
rover             1586.492308
saab              3183.493671
seat              4320.168662
skoda             6353.544872
smart             3518.102305
sonstige_a

In [71]:
#Step2: Mileage 

#With this method, the keys in the dictionary became the index in the series object

M = pd.Series(mean_mileage)

print(M)

alfa_romeo       131043.613707
audi             129492.562380
bmw              132686.255924
chevrolet         99472.727273
chrysler         132366.863905
citroen          120043.731778
dacia             84728.682171
daewoo           121644.736842
daihatsu         115901.639344
fiat             117593.032462
ford             124327.718676
honda            123092.783505
hyundai          106718.426501
jaguar           124527.027027
jeep             127102.803738
kia              112739.130435
lada              85000.000000
lancia           123090.909091
land_rover       118333.333333
mazda            124871.447903
mercedes_benz    130848.387097
mini              88899.521531
mitsubishi       126549.118388
nissan           118711.201080
opel             129455.924171
peugeot          127356.643357
porsche           96759.581882
renault          128062.365591
rover            138230.769231
saab             143670.886076
seat             121768.226333
skoda            110993.589744
smart   

In [78]:

#Step 2: 
#We can then create a two-column dataframe from both the series objects.

#create a dictionary
d={'Mean_price': P, "Mean_mileage":M}

df = pd.DataFrame(data=d)

df

Mean_mileage   Mean_price
alfa_romeo     131043.613707  4041.844237
audi           129492.562380  9212.930662
bmw            132686.255924  8493.857251
chevrolet       99472.727273  6620.069091
chrysler       132366.863905  3519.384615
citroen        120043.731778 43578.909621
dacia           84728.682171  5897.736434
daewoo         121644.736842  1079.342105
daihatsu       115901.639344  1628.426230
fiat           117593.032462 12566.541568
ford           124327.718676  7305.141844
honda          123092.783505  4000.139175
hyundai        106718.426501  5371.792961
jaguar         124527.027027 11525.554054
jeep           127102.803738 11590.214953
kia            112739.130435  5889.298551
lada            85000.000000  2647.724138
lancia         123090.909091  3182.000000
land_rover     118333.333333 18934.272727
mazda          124871.447903  4059.059540
mercedes_benz  130848.387097  8526.623226
mini            88899.521531 10541.566986
mitsubishi     126549.118388  3386.911839
nissan         118711.201080  4669.385965
opel           129455.924171  5281.321706
peugeot        127356.643357  3065.611888
porsche         96759.581882 45624.271777
renault        128062.365591  2431.195699
rover          138230.769231  1586.492308
saab           143670.886076  3183.493671
seat           121768.226333  4320.168662
skoda          110993.589744  6353.544872
smart          100511.527378  3518.102305
sonstige_autos  90467.091295 44381.441614
subaru         125784.313725  3940.470588
suzuki         108776.223776  4137.807692
toyota         116219.312602  5148.003273
trabant         55149.253731  1789.626866
volkswagen     128961.679892  6600.382233
volvo          138363.636364 32914.290909

### Step 7: Next considerations

Data cleaning next steps:
- Identify categorical data that uses german words, translate them and map the values to their english counterparts
- Convert the dates to be uniform numeric data, so "2016-03-21" becomes the integer 20160321.
- See if there are particular keywords in the name column that you can extract as new columns

#### Analysis next steps:
1. Find the most common brand/model combinations
2. Split the odometer_km into groups, and use aggregation to see if average prices follows any patterns based on the milage.
3. How much cheaper are cars with damage than their non-damaged counterparts?